### PR TITLE
Initial implementation of new filter and output API.

### PR DESCRIPTION
**Pull request by rafrombrc**

### CODE REVIEW

**Comment by rafrombrc** [comment]
Work in progress... documentation is still coming, just getting this up so code review can start while docs are being written.

File: pipeline/plugin_runners.go [hunk 1]
```diff
@@ -769,21 +794,30 @@ func (kind foRunnerKind) String() string {
...
+	useFraming          bool    // output only
+	canExit             bool
+	useBuffering        bool
+	kind                foRunnerKind
+	pConfig             *PipelineConfig
+	lastErr             error
+	bufReader           *BufferReader
+	stopChan            chan bool
+	processMessageCount int64
+	dropMessageCount    int64
```
Comment (trink) [inline pipeline/plugin_runners.go]: Move these two items to the top of the struct to avoid alignment issues on 32 bit systems

**Comment by rafrombrc** [comment]
Travis seems to be failing randomly at this point, tests are consistently passing locally. This is ready for final review.

File: sandbox/plugins/sandbox_output.go [hunk 2]
```diff
@@ -125,18 +126,23 @@ func (s *SandboxOutput) Run(or pipeline.OutputRunner, h pipeline.PluginHelper) (
 				s.reportLock.Unlock()
 			}
 			s.sample = 0 == rand.Intn(s.sampleDenominator)
-			pack.Recycle()
+			// pack.Recycle(nil)
```
Comment (trink) [inline sandbox/plugins/sandbox_output.go]: remove

File: sandbox/plugins/sandbox_filter.go [hunk 3]
```diff
@@ -180,9 +181,14 @@ func (this *SandboxFilter) Run(fr pipeline.FilterRunner, h pipeline.PluginHelper
 		startTime      time.Time
 		slowDuration   int64 = int64(this.pConfig.Globals.MaxMsgProcessDuration)
 		duration       int64
-		capacity       = cap(inChan) - 1
+		samplesNeeded  int64
 	)
 
+	if fr.UsesBuffering() {
+		samplesNeeded = 29
```
Comment (trink) [inline sandbox/plugins/sandbox_filter.go]: Not ideal hard coding it to one less than the default router channel size.

File: plugins/tcp/tcp_output_test.go [hunk 4]
```diff
@@ -209,91 +194,99 @@ func TcpOutputSpec(c gs.Context) {
...
 
-			c.Expect(atomic.LoadInt64(&tcpOutput.dropMessageCount), gs.Equals, int64(0))
-			c.Expect(atomic.LoadInt64(&tcpOutput.processMessageCount), gs.Equals, int64(0))
+		// 	c.Expect(atomic.LoadInt64(&tcpOutput.dropMessageCount), gs.Equals, int64(0))
+		// 	c.Expect(atomic.LoadInt64(&tcpOutput.processMessageCount), gs.Equals, int64(0))
 
-			close(inChan)
-		})
+		// 	close(inChan)
+		// })
```
Comment (trink) [inline plugins/tcp/tcp_output_test.go]: Remove old commented out test fragments.

**Comment by trink** [comment]
Not sure if you want to remove the old commented out code or make the samplesNeeded match the plugin_chansize configuration - 1 but  I am ok with getting this merged and having people put some bake time on it.

**Comment by rafrombrc** [comment]
Great, thanks... I've found some issues in my own testing w/ the ElasticSearchOutput, working on those now, will get all this resolved and merged ASAP.

**Pull request merged**